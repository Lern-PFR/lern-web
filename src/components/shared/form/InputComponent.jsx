import styled from 'styled-components';
import {
	border,
	display,
	layout,
	shadow,
	color,
} from 'styled-system';
import PropTypes from 'prop-types';
import { disabled as disabledStyle } from 'theme/buttonStyles';
import { inputStyle, errorInputStyle } from 'theme/formStyles';
import { forwardRef, useRef } from 'react';

const StyledInputComponent = styled('input')(
	{ ...inputStyle },
	border,
	display,
	layout,
	shadow,
	color,
);

/**
 * @name InputComponent
 * @description A component used to display styled input element.
 *
 * @author Christopher Walker
 *
 * @param {bool}	[disabled]			: Whether the input is disabled.
 * @param {string}	[type]				: The type of input, default is text.
 * @param {string}	[placeholder]		: Placeholder text for this input.
 * @param {bool}	[hasError]			: Whether the input contains a validation error.
 * @param {string}	id					: The id of the input.
 */

const InputComponent = forwardRef(
	({ id, disabled, placeholder, type, hasError, ...otherProps }, ref) => {
		const defaultRef = useRef();
		const resolvedRef = ref || defaultRef;

		return (
			<StyledInputComponent
				{...(disabled ? disabledStyle : {})}
				type={type}
				disabled={disabled}
				placeholder={placeholder}
				id={id}
				ref={resolvedRef}
				{...(hasError ? errorInputStyle : {})}
				{...otherProps}
			/>
		);
	}
);

InputComponent.displayName = 'InputComponent';

InputComponent.propTypes = {
	id: PropTypes.string.isRequired,
	placeholder: PropTypes.string,
	disabled: PropTypes.bool,
	hasError: PropTypes.bool,
	type: PropTypes.string,
};

InputComponent.defaultProps = {
	disabled: false,
	hasError: false,
	type: 'text',
	placeholder: '',
};

export default InputComponent;
