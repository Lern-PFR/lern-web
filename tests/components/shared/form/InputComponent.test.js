import { shallow } from 'enzyme';
import InputComponent from 'components/shared/form/InputComponent';
import { cleanup } from '@testing-library/react';

describe('RadioButtonComponent', () => {
	afterEach(() => {
		cleanup();
	});

	it('should match previous snapshot', () => {
		const sut = (<InputComponent id="inp" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for disabled state', () => {
		const sut = (<InputComponent disabled id="inp" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for input with custom placeholder', () => {
		const sut = (<InputComponent id="inp" placeholder="Enter details here" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for input with validation error', () => {
		const sut = (<InputComponent id="inp" hasError />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for input with specified type', () => {
		const sut = (<InputComponent id="inp" type="password" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
