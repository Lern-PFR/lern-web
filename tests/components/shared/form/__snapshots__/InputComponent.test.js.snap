// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonComponent should match previous snapshot 1`] = `
<styled.input
  disabled={false}
  id="inp"
  placeholder="Type here."
  type="text"
/>
`;

exports[`RadioButtonComponent should match previous snapshot for disabled state 1`] = `
<styled.input
  bg="F6F6F4"
  color="#8C8C8C"
  disabled={true}
  id="inp"
  placeholder="Type here."
  type="text"
/>
`;

exports[`RadioButtonComponent should match previous snapshot for input with custom placeholder 1`] = `
<styled.input
  disabled={false}
  id="inp"
  placeholder="Enter details here"
  type="text"
/>
`;

exports[`RadioButtonComponent should match previous snapshot for input with specified type 1`] = `
<styled.input
  disabled={false}
  id="inp"
  placeholder="Type here."
  type="password"
/>
`;

exports[`RadioButtonComponent should match previous snapshot for input with validation error 1`] = `
<styled.input
  border="solid 1px red"
  disabled={false}
  id="inp"
  placeholder="Type here."
  type="text"
/>
`;
